onnxruntime/core/session/inference_session.cc

```diff
@@ -2299,6 +2299,112 @@ Status InferenceSession::Run(const RunOptions& run_options,
+    if (input_names[i] == nullptr || input_names[i][0] == '\0') {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "input name cannot be empty");
+    }
+
+    if (!input[i]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, MakeString("NULL input supplied for input ", input_names[i]).c_str());
+    }
+
+    feed_names.emplace_back(input_names[i]);
+    feeds.emplace_back(*input[i]);
+  }
+
+  // Create output feed
+  InlinedVector<std::string> output_name_vec;
+  output_name_vec.reserve(output_names_len);
+  for (size_t i = 0; i != output_names_len; ++i) {
+    if (output_names[i] == nullptr || output_names[i][0] == '\0') {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "output name cannot be empty");
+    }
+    output_name_vec.emplace_back(output_names[i]);
+  }
+
+  std::vector<OrtValue> fetches;
+  fetches.reserve(output_names_len);
+  for (size_t i = 0; i != output_names_len; ++i) {
+    if (output[i] != nullptr) {
+      fetches.emplace_back(*output[i]);
+    } else {
+      fetches.emplace_back();
+    }
+  }
+
+  Status status;
+  if (run_options == nullptr) {
+    OrtRunOptions op;
+    status = Run(op, feed_names, feeds, output_name_vec, &fetches, nullptr);
+  } else {
+    status = Run(*run_options, feed_names, feeds, output_name_vec, &fetches, nullptr);
+  }
+
+  if (!status.IsOK())
+    return status;
+
+  // We do it in two loops to make sure copy __ctors does not throw
+  InlinedVector<std::unique_ptr<OrtValue>> output_unique_ptrs;
+  output_unique_ptrs.reserve(output_names_len);
+  for (size_t i = 0; i != output_names_len; ++i) {
+    if (output[i] == nullptr) {
+      output_unique_ptrs.emplace_back(std::make_unique<OrtValue>(fetches[i]));
+    } else {
+      output_unique_ptrs.emplace_back();
+    }
+  }
+
+  ORT_ENFORCE(output_unique_ptrs.size() == output_names_len);
+
+  for (size_t i = 0; i != output_names_len; ++i) {
+    if (output[i] == nullptr) {
+      ORT_ENFORCE(output_unique_ptrs[i] != nullptr);
+      output[i] = output_unique_ptrs[i].release();
+    }
+  }
+  return Status::OK();
+}
+
+Status InferenceSession::RunAsync(const OrtRunOptions* run_options, const char* const* input_names,
+                                  const OrtValue* const* input, size_t input_len,
+                                  const char* const* output_name, size_t output_names_len,
+                                  RunAsyncCallbackFn callback, void* user_data) {
+  InferenceSession* sess = this;
+  std::function<void()> run_fn = [=]() {
+    ORT_TRY {
+      using OrtValuePtr = OrtValue*;
+      std::unique_ptr<OrtValuePtr[]> outputs = std::make_unique<OrtValuePtr[]>(output_names_len);
+      memset(outputs.get(), 0x0, sizeof(OrtValuePtr) * output_names_len);
+      auto status = sess->Run(run_options, input_names, input, input_len, output_name, output_names_len, outputs.get());
+      if (status.IsOK()) {
+        callback(user_data, outputs.get(), output_names_len, {});
```

Comment:
This is dangerous. Calling get() will simply return the underlying ptr. When outputs is freed at the end of this function, these underlying ptrs given the user will be dangling. We should use release(). See https://github.com/microsoft/onnxruntime/blob/dd72192cf4866e7ca036a7987a5ef76f930ded3a/onnxruntime/core/session/onnxruntime_c_api.cc#L888-L893

Reply:
Yes, the Release() approach was considered and tested.
However, it still does not solve the question - "who would release each **element** of the output array?"
Worse than that - it introduces another burden to the customer - they will have to release the output array, plus each of its elements.

Reply:
It's based on design.

If we expect users to use the `outputs` only ( and they don't take the ownership ), as long as the callback is a synchronized function, it is fine.
 - in this design, using `std::vector<OrtValuePtr>` is better than using `std::unique_ptr<OrtValuePtr[]>`.

If we expect users to take the ownership of `outputs` and be responsible to release them, we should use `.release()` here. In this case, we should use C-API to allocate the memory so that users can use the corrsponding C-API to release them. 
 - in this design, should use a set of `OrtAllocator*` APIs to allocate memory for those `OrtValuePtr`s and expect users to call `AllocatorFree()` and `OrtReleaseValue()` for each outputs, and, if necessary, also pass the `AllocatorPtr`.

since the second design heavily depends on users' code to do clean up, I recommend to use the first design.